app/process_proposal.go

```diff
@@ -25,6 +25,11 @@ const rejectedPropBlockLog = "Rejected proposal block:"
 
 func (app *App) ProcessProposalHandler(ctx sdk.Context, req *abci.RequestProcessProposal) (resp *abci.ResponseProcessProposal, err error) {
 	defer telemetry.MeasureSince(time.Now(), "process_proposal")
+
+	// Clear the cache after ProcessProposal completes (whether accept or reject)
+	// to prevent memory growth and ensure fresh cache for the next block proposal
+	defer app.txValidationCache.Clear()
```

Comment:
I don't think we should be clearing in process proposal. If a block makes it to process proposal, it doesn't mean it will be finalized. So I think we should clear the cache in the finalizeBlock and only for transactions that were included.